File: output_log/claude-opus/sample_1/c/qwen3/c/Struct/pred.c
#include "stdlib.h"

/*@ 
predicate account(struct account *p; int limit, int balance) =
    p != 0 &*&
    p->limit |-> limit &*&
    p->balance |-> balance &*&
    limit <= balance;
@*/

struct account
{
    int limit;
    int balance;
};

struct account *create_account(int limit)
    //@ requires true;
    //@ ensures account(result, limit, 0);
{
    struct account *myAccount = malloc(sizeof(struct account));
    if (myAccount == 0)
    {
        abort();
    }
    //@ assume limit <= 0;
    myAccount->limit = limit;
    myAccount->balance = 0;

    //@ assert limit <= 0;
    //@ assert limit <= 0 + 0;
    //@ close account(myAccount, limit, 0);
    return myAccount;
}

int account_get_balance(struct account *myAccount)
    //@ requires account(myAccount, ?limit, ?balance);
    //@ ensures account(myAccount, limit, balance) &*& result == balance;
{
    int result = myAccount->balance;

    return result;
}

void account_deposit(struct account *myAccount, int amount)
    //@ requires account(myAccount, ?limit, ?balance);
    //@ ensures account(myAccount, limit, balance + amount);
{
    int old_balance = myAccount->balance;
    //@ open account(myAccount, limit, old_balance);
    myAccount->balance += amount;
    //@ close account(myAccount, limit, old_balance + amount);
}

int account_withdraw(struct account *myAccount, int amount)
    //@ requires account(myAccount, ?limit, ?balance);
    //@ ensures account(myAccount, limit, balance - result) &*&
    //@         (balance - amount < limit ==> result == balance - limit) &*&
    //@         (balance - amount >= limit ==> result == amount);
{
    //@ open account(myAccount, limit, balance);
    int to_withdraw;
    //@ int can_withdraw_minus = balance - amount;
    //@ int cond = 0; // dummy variable because conditional expressions forbidden in specs
    //@ // We cannot write cond as (balance - amount < limit) because that would be conditional
    if (myAccount->balance - amount < myAccount->limit) {
        to_withdraw = myAccount->balance - myAccount->limit;
        //@ //@ under this branch: result == balance - limit 
    } else {
        to_withdraw = amount;
        //@ //@ under this branch: result == amount
    }
    myAccount->balance -= to_withdraw;
    //@ close account(myAccount, limit, balance - to_withdraw);
    return to_withdraw;
}

void account_dispose(struct account *myAccount)
    //@ requires account(myAccount, ?limit, ?balance);
    //@ ensures true;
{
    //@ open account(myAccount, limit, balance);
    free(myAccount);
}

int main()
    //@ requires true;
    //@ ensures true;
{
    struct account *myAccount = create_account(-100);
    account_deposit(myAccount, 200);
    int w1 = account_withdraw(myAccount, 50);
    //@ assert w1 == 50;
    int b1 = account_get_balance(myAccount);
    //@ assert b1 == 150;
    int w2 = account_withdraw(myAccount, 300);
    //@ assert w2 == 250;
    int b2 = account_get_balance(myAccount);
    //@ assert b2 == -100;
    account_dispose(myAccount);
    return 0;
}

File: output_at_5/claude-opus/sample_2/rust/rust/Struct/pred.rs
use std::alloc::{Layout, alloc, handle_alloc_error, dealloc};

struct Account {
    limit: i32,
    balance: i32,
}

//@ /*
//@ pred account(Account* a; int limit, int balance) =
//@     a->limit |-> limit &*& a->balance |-> balance;
//@ */

impl Account {

    unsafe fn create(limit: i32) -> *mut Account
    //@ requires true;
    //@ ensures result != 0 && account(result, limit, 0);
    {
        let my_account = alloc(Layout::new::<Account>()) as *mut Account;
        if my_account.is_null() {
            handle_alloc_error(Layout::new::<Account>());
        }
        (*my_account).limit = limit;
        (*my_account).balance = 0;
        //@ close account(my_account, limit, 0);
        my_account
    }

    unsafe fn get_balance(my_account: *mut Account) -> i32
    //@ requires account(my_account, ?limit, ?balance);
    //@ ensures account(my_account, limit, balance) &*& result == balance;
    {
        //@ open account(my_account, limit, balance);
        let result = (*my_account).balance;
        //@ close account(my_account, limit, balance);
        result
    }

    unsafe fn deposit(my_account: *mut Account, amount: i32)
    //@ requires account(my_account, ?limit, ?balance);
    //@ ensures account(my_account, limit, balance + amount);
    {
        //@ open account(my_account, limit, balance);
        (*my_account).balance += amount;
        //@ close account(my_account, limit, balance + amount);
    }

    unsafe fn withdraw(my_account: *mut Account, amount: i32) -> i32
    //@ requires account(my_account, ?limit, ?balance);
    //@ ensures account(my_account, limit, balance - result) &*&
    //@         (((balance - amount) < limit && result == balance - limit) || ((balance - amount) >= limit && result == amount));
    {
        //@ open account(my_account, limit, balance);
        let result =
            if (*my_account).balance - amount < (*my_account).limit {
                (*my_account).balance - (*my_account).limit
            } else {
                amount
            };
        (*my_account).balance -= result;
        //@ close account(my_account, limit, balance - result);
        result
    }

    unsafe fn dispose(my_account: *mut Account)
    //@ requires account(my_account, _, _);
    //@ ensures true;
    {
        //@ open account(my_account, _, _);
        dealloc(my_account as *mut u8, Layout::new::<Account>());
    }

}

fn main()
{
    unsafe {
        let my_account = Account::create(-100);
        Account::deposit(my_account, 200);
        let w1 = Account::withdraw(my_account, 50);
        std::hint::assert_unchecked(w1 == 50);
        let b1 = Account::get_balance(my_account);
        std::hint::assert_unchecked(b1 == 150);
        let w2 = Account::withdraw(my_account, 300);
        std::hint::assert_unchecked(w2 == 250);
        let b2 = Account::get_balance(my_account);
        std::hint::assert_unchecked(b2 == -100);
        Account::dispose(my_account);
    }
}